File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let translate = function
    | "AUG" -> "Methionine"
    | "UUU" | "UUC" -> "Phenylalanine"
    | "UUA" | "UUG" -> "Leucine"
    | "UCU" | "UCC" | "UCA" | "UCG" -> "Serine"
    | "UAU" | "UAC" -> "Tyrosine"
    | "UGU" | "UGC" -> "Cysteine"
    | "UGG" -> "Tryptophan"
    | "UAA" | "UAG" | "UGA" -> "STOP"
    | codon -> failwith $"Unknown codon {codon}"

let proteins rna =
    Seq.chunkBySize 3 rna
    |> Seq.map (System.String >> translate)
    |> Seq.takeWhile ((<>) "STOP")
    |> Seq.toList

File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;

public static class ProteinTranslation
{
    public static Dictionary<string, string> CodonToProtein { get; private set; }

     static ProteinTranslation() {
         CodonToProtein = new Dictionary<string, string>
         {
             { "AUG", "Methionine" }, { "UUU", "Phenylalanine" },
             { "UUC", "Phenylalanine" }, { "UUA", "Leucine" },
             { "UUG", "Leucine" }, { "UCU", "Serine" },
             { "UCC", "Serine" }, { "UCA", "Serine" },
             { "UCG", "Serine" },  { "UAU", "Tyrosine" },
             { "UAC", "Tyrosine" }, { "UGU", "Cysteine" },
             { "UGC", "Cysteine" }, { "UGG", "Tryptophan" },
             { "UAA", "STOP" }, { "UAG", "STOP" }, { "UGA", "STOP" }
         };
     }
    
    public static string[] Proteins(string strand)
    {
        List<string> proteins = new List<string> { };
        for (int i = 0; i < strand.Length; i = i + 3)
        {
            string codon = strand.Substring(i, 3);
            string protein = ProteinTranslation.CodonToProtein[codon];
            if (protein == "STOP") break;
            proteins.Add(protein);
        }
        return proteins.ToArray();
    }
}